### PR TITLE
Core: Use ParallelIterable in Deletes::toPositionIndex (6387)

### PR DESCRIPTION
**Pull request by wypoon**
This is a continuation of https://github.com/apache/iceberg/pull/6432 by @rbalamohan, as he is not working on it anymore.
I have squashed all his commits in that PR into one, rebased it on master, and made some minor changes.

### CODE REVIEW

**Comment by wypoon** [comment]
@rdblue @aokolnychyi as @rbalamohan indicated that he's not working on https://github.com/apache/iceberg/pull/6432 anymore, I have taken it up here. I rebased it on master and resolved the conflicts, moving the configuration from `SystemProperties` to the new `SystemConfigs`, changed the default for the pool size to be the same as for the existing worker pool, and the tests are green.
@aokolnychyi I didn't see https://github.com/apache/iceberg/pull/6432#issuecomment-1758777424 until I have put this up. I didn't know that you're also working on this in https://github.com/apache/iceberg/pull/8755. I'll take a look at that.

File: core/src/test/java/org/apache/iceberg/deletes/TestPositionFilter.java [hunk 1]
```diff
@@ -282,6 +286,16 @@ public void testPositionSetRowFilter() {
 
   @Test
   public void testCombinedPositionSetRowFilter() {
+    testCombinedPositionSetRowFilter(null);
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/deletes/TestPositionFilter.java:289]: minor: why not make the test a `@ParameterizedTest`, taking a `null` and a configured `ExecutorService`?

Comment (wypoon) [reply core/src/test/java/org/apache/iceberg/deletes/TestPositionFilter.java]: Done.

File: core/src/main/java/org/apache/iceberg/deletes/Deletes.java [hunk 2]
```diff
@@ -127,14 +130,25 @@ public static StructLikeSet toEqualitySet(
...
+      ExecutorService deleteWorkerPool) {
     DataFileFilter<T> locationFilter = new DataFileFilter<>(dataLocation);
     List<CloseableIterable<Long>> positions =
         Lists.transform(
             deleteFiles,
             deletes ->
                 CloseableIterable.transform(
                     locationFilter.filter(deletes), row -> (Long) POSITION_ACCESSOR.get(row)));
-    return toPositionIndex(CloseableIterable.concat(positions));
+    if (positions.size() > 1 && (deleteWorkerPool != null)) {
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/deletes/Deletes.java:147]:
```suggestion
    if (positions.size() > 1 && deleteWorkerPool != null) {
```

File: core/src/test/java/org/apache/iceberg/deletes/TestPositionFilter.java [hunk 3]
```diff
@@ -280,8 +287,16 @@ public void testPositionSetRowFilter() {
         .containsExactlyElementsOf(Lists.newArrayList(1L, 2L, 5L, 6L, 8L));
   }
 
-  @Test
-  public void testCombinedPositionSetRowFilter() {
+  static Stream<ExecutorService> executorServiceProvider() {
+    ExecutorService executorService =
```
Comment (nastra) [inline core/src/test/java/org/apache/iceberg/deletes/TestPositionFilter.java:291]: nit: maybe this can be inlined?

**Review by nastra — APPROVED** [review]
LGTM with some nits

File: core/src/main/java/org/apache/iceberg/deletes/Deletes.java (hunk 2 above)
Comment (wypoon) [reply core/src/main/java/org/apache/iceberg/deletes/Deletes.java]: Done.

File: core/src/test/java/org/apache/iceberg/deletes/TestPositionFilter.java (hunk 3 above)
Comment (wypoon) [reply core/src/test/java/org/apache/iceberg/deletes/TestPositionFilter.java]: Done.

**Comment by nastra** [comment]
LGTM, thanks @wypoon

**Comment by aokolnychyi** [comment]
I actually have some concerns about this change, which I outlined in [this](https://docs.google.com/document/d/1M4L6o-qnGRwGhbhkW8BnravoTwvCrJV8VvzVQDRJO5I) doc.
I have an alternative implementation in PR #6432 that I want to consider before this change is released.

**Pull request merged**